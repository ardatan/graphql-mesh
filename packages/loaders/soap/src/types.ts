export interface WSDLDefinitionAttributes {
  name: string;
  targetNamespace: string;
}

export interface WSDLDefinition {
  attributes: WSDLDefinitionAttributes;
  import?: WSDLImport[];
  service?: WSDLService[];
  binding?: WSDLBinding[];
  types?: WSDLTypes[];
  message?: WSDLMessage[];
  portType?: WSDLPortType[];
}

export interface WSDLPortTypeAttributes {
  name: string;
}

export interface WSDLPortType {
  attributes: WSDLPortTypeAttributes;
  operation: WSDLOperation[];
}

export interface WSDLMessageAttributes {
  name: string;
}

export interface WSDLPartAttributes {
  element: string;
  type: string;
  name: string;
}

export interface WSDLPart {
  attributes: WSDLPartAttributes;
}

export interface WSDLMessage {
  attributes: WSDLMessageAttributes;
  part: WSDLPart[];
}

export interface XSSchemaAttributes {
  id: string;
  targetNamespace: string;
  version: string;
}

export interface XSSchema {
  attributes: XSSchemaAttributes;
  element?: XSElement[];
  complexType?: XSComplexType[];
  simpleType?: XSSimpleType[];
  import?: XSImport[];
}

export interface XSImport {
  attributes: XSImportAttributes;
}

export interface XSImportAttributes {
  namespace: string;
  schemaLocation: string;
}

export interface XSSimpleTypeAttributes {
  name: string;
}

export interface XSSimpleType {
  attributes: XSSimpleTypeAttributes;
  restriction: XSRestriction[];
}

export interface XSRestrictionAttributes {
  base: string;
}

export interface XSRestriction {
  pattern: XSPattern[];
  attributes: XSRestrictionAttributes;
  enumeration: XSEnumeration[];
}

export interface XSPattern {
  attributes: XSPatternAttributes;
}

export interface XSPatternAttributes {
  value: string;
}

export interface XSEnumerationAttributes {
  value: string;
}

export interface XSEnumeration {
  attributes: XSEnumerationAttributes;
}

export interface XSComplexTypeAttributes {
  name: string;
}

export interface XSElementAttributes {
  name: string;
  type: string;
  maxOccurs?: string;
  minOccurs?: string;
  nillable?: 'true' | 'false';
  ref?: string;
}

export interface XSElement {
  attributes: XSElementAttributes;
  complexType?: XSComplexType[];
  simpleType?: XSSimpleType[];
}

export interface XSSequenceAttributes {
  maxOccurs?: string;
  minOccurs?: string;
  nillable?: 'true' | 'false';
}

export interface XSSequence {
  attributes?: XSSequenceAttributes;
  element?: XSElement[];
  any?: XSAny[];
}

export interface XSChoiceAttributes {
  maxOccurs?: string;
  minOccurs?: string;
  nillable?: 'true' | 'false';
}

export interface XSChoice {
  attributes?: XSChoiceAttributes;
  element?: XSElement[];
  any?: XSAny[];
}

export interface XSAny {
  attributes?: XSAnyAttributes;
}

export interface XSAnyAttributes {
  namespace?: string;
}

export interface XSExtensionAttributes {
  base: string;
}

export interface XSExtension {
  attributes: XSExtensionAttributes;
  complexContent?: XSComplexContent[];
  sequence?: XSSequence[];
  choice?: XSChoice[];
}

export interface XSComplexContent {
  extension: XSExtension[];
}

export interface XSComplexType {
  attributes?: XSComplexTypeAttributes;
  complexContent?: XSComplexContent[];
  sequence?: XSSequence[];
  choice?: XSChoice[];
}

export interface WSDLTypes {
  schema: XSSchema[];
}

export interface WSDLBindingAttributes {
  name: string;
  type: string;
}

export interface SOAPBindingAttributes {
  style: string;
  transport: string;
}

export interface SOAPBinding {
  attributes: SOAPBindingAttributes;
}

export interface WSDLBinding {
  attributes: WSDLBindingAttributes;
  binding: SOAPBinding;
  operation: WSDLOperation[];
}

export interface WSDLOperationAttributes {
  name: string;
}

export interface SOAPOperationAttributes {
  soapAction: string;
  style: string;
}

export interface SOAPOperation {
  attributes: SOAPOperationAttributes;
}

export interface WSDLInputAttributes {
  name: string;
  message: string;
}

export interface SOAPBodyAttributes {
  use: string;
}

export interface SOAPBody {
  attributes: SOAPBodyAttributes;
}

export interface WSDLInput {
  attributes: WSDLInputAttributes;
  body: SOAPBody[];
}

export interface WSDLOutputAttributes {
  name: string;
  message: string;
}

export interface WSDLOutput {
  attributes: WSDLOutputAttributes;
  body: SOAPBody[];
}

export interface WSDLFaultAttributes {
  name: string;
}

export interface SOAPFault {
  attributes: SOAPFaultAttributes;
}

export interface SOAPFaultAttributes {
  name: string;
  use: string;
}

export interface WSDLFault {
  attributes: WSDLFaultAttributes;
  fault: SOAPFault[];
}

export interface WSDLOperation {
  attributes: WSDLOperationAttributes;
  operation?: SOAPOperation[];
  input: WSDLInput[];
  output: WSDLOutput[];
  fault: WSDLFault[];
}

export interface WSDLServiceAttributes {
  name: string;
}

export interface WSDLService {
  attributes: WSDLServiceAttributes;
  port: WSDLPort[];
}

export interface WSDLPortAttributes {
  binding: string;
  name: string;
}

export interface WSDLPort {
  attributes: WSDLPortAttributes;
  address: WSDLAddress[];
}

export interface WSDLAddressAttributes {
  location: string;
}

export interface WSDLAddress {
  attributes: WSDLAddressAttributes;
}

export interface WSDLImportAttributes {
  location?: string;
  namespace: string;
}

export interface WSDLImport {
  attributes: WSDLImportAttributes;
}

export interface WSDLObject {
  definitions: WSDLDefinition[];
}

export interface XSDObject {
  schema: XSSchema[];
}
