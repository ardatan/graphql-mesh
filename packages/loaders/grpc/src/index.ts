import type { Logger, YamlConfig } from "@graphql-mesh/types";
import { gRPCLoader } from "./gRPCLoader";

interface LoaderContext {
  cwd: string;
  logger: Logger;
}

export function loadGrpcSubgraph(name: string, options: YamlConfig.GrpcHandler) {
  return (ctx: LoaderContext) => ({
    name,
    schema$: new gRPCLoader(name, ctx.cwd, ctx.logger, options).buildSchema(),
  });
}
