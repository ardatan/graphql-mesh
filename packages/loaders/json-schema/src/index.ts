import { loadGraphQLSchemaFromJSONSchemas } from './loadGraphQLSchemaFromJSONSchemas.js';
export default loadGraphQLSchemaFromJSONSchemas;
export * from './loadGraphQLSchemaFromJSONSchemas.js';
export * from './getComposerFromJSONSchema.js';
export * from './getDereferencedJSONSchemaFromOperations.js';
export * from './getGraphQLSchemaFromDereferencedJSONSchema.js';
export * from './types.js';
export { processDirectives } from './directives.js';
export * from './bundle.js';
