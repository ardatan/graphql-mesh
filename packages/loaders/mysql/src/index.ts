import { GraphQLSchema } from 'graphql';
import { Logger, MeshPubSub } from '@graphql-mesh/types';
import { getMySQLExecutor } from './execution.js';
import { loadGraphQLSchemaFromMySQL, LoadGraphQLSchemaFromMySQLOpts } from './schema.js';

export * from './schema.js';
export * from './execution.js';

export function loadMySQLSubgraph(name: string, opts: LoadGraphQLSchemaFromMySQLOpts) {
  return () => ({
    name,
    schema$: loadGraphQLSchemaFromMySQL(name, opts),
  });
}

export interface MySQLTransportEntry {
  kind: 'mysql';
  location: string;
}

export function getSubgraphExecutor(transportContext: {
  getSubgraph: () => GraphQLSchema;
  pubsub: MeshPubSub;
  logger: Logger;
}) {
  return getMySQLExecutor({
    subgraph: transportContext.getSubgraph(),
    pubsub: transportContext.pubsub,
    logger: transportContext.logger,
  });
}
