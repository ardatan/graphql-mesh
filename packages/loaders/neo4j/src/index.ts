import type { Logger } from '@graphql-mesh/types';
import type { LoadGraphQLSchemaFromNeo4JOpts } from './schema.js';
import { loadGraphQLSchemaFromNeo4J } from './schema.js';

export function loadNeo4JSubgraph(name: string, opts: LoadGraphQLSchemaFromNeo4JOpts) {
  return ({ logger }: { logger: Logger }) => ({
    name,
    schema$: loadGraphQLSchemaFromNeo4J(name, {
      ...opts,
      logger,
    }),
  });
}

export * from './schema.js';
export * from '@graphql-mesh/transport-neo4j';
