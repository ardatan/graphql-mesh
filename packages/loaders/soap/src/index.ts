export * from './SOAPLoader';
export * from './types';
export * from './executor';
