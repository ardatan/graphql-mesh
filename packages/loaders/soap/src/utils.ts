import { X2jOptions } from 'fast-xml-parser';

export interface SoapAnnotations {
  elementName: string;
  bindingNamespace: string;
  endpoint: string;
}

export const PARSE_XML_OPTIONS: Partial<X2jOptions> = {
  attributeNamePrefix: '',
  attrNodeName: 'attributes',
  textNodeName: 'innerText',
  ignoreAttributes: false,
  ignoreNameSpace: true,
  arrayMode: true,
  allowBooleanAttributes: true,
};
