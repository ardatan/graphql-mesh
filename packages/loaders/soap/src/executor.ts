import {
  execute,
  GraphQLFieldResolver,
  GraphQLOutputType,
  GraphQLResolveInfo,
  GraphQLSchema,
  isListType,
  isNonNullType,
} from 'graphql';
import { parse as parseXML, j2xParser as JSONToXMLConverter } from 'fast-xml-parser';
import { MeshFetch } from '@graphql-mesh/types';
import { PARSE_XML_OPTIONS, SoapAnnotations } from './utils.js';
import { Executor, getDirective, getRootTypes } from '@graphql-tools/utils';

function isOriginallyListType(type: GraphQLOutputType): boolean {
  if (isNonNullType(type)) {
    return isOriginallyListType(type.ofType);
  }
  return isListType(type);
}

const defaultFieldResolver: GraphQLFieldResolver<any, any> = function soapDefaultResolver(root, args, context, info) {
  const rootField = root[info.fieldName];
  if (typeof rootField === 'function') {
    return rootField(args, context, info);
  }
  const fieldValue = rootField;
  const isArray = Array.isArray(fieldValue);
  const isPlural = isOriginallyListType(info.returnType);
  if (isPlural && !isArray) {
    return [fieldValue];
  }
  if (!isPlural && isArray) {
    return fieldValue[0];
  }
  return fieldValue;
};

function normalizeArgsForConverter(args: any): any {
  if (args != null) {
    if (typeof args === 'object') {
      for (const key in args) {
        args[key] = normalizeArgsForConverter(args[key]);
      }
    } else {
      return {
        innerText: args,
      };
    }
  }
  return args;
}

function normalizeResult(result: any) {
  if (result != null && typeof result === 'object') {
    for (const key in result) {
      if (key === 'innerText') {
        return result.innerText;
      }
      result[key] = normalizeResult(result[key]);
    }
    if (Array.isArray(result) && result.length === 1) {
      return result[0];
    }
  }
  return result;
}

type RootValueMethod = (args: any, context: any, info: GraphQLResolveInfo) => Promise<any>;

function createRootValueMethod(soapAnnotations: SoapAnnotations, fetchFn: MeshFetch): RootValueMethod {
  const jsonToXMLConverter = new JSONToXMLConverter({
    attributeNamePrefix: '',
    attrNodeName: 'attributes',
    textNodeName: 'innerText',
  });

  return async function rootValueMethod(args: any, context: any, info: GraphQLResolveInfo) {
    const requestJson = {
      'soap:Envelope': {
        attributes: {
          'xmlns:soap': 'http://www.w3.org/2003/05/soap-envelope',
        },
        'soap:Body': {
          attributes: {
            xmlns: soapAnnotations.bindingNamespace,
          },
          ...normalizeArgsForConverter(args),
        },
      },
    };
    const requestXML = jsonToXMLConverter.parse(requestJson);
    const response = await fetchFn(
      soapAnnotations.baseUrl,
      {
        method: 'POST',
        body: requestXML,
        headers: {
          'Content-Type': 'application/soap+xml; charset=utf-8',
        },
      },
      context,
      info
    );
    const responseXML = await response.text();
    const responseJSON = parseXML(responseXML, PARSE_XML_OPTIONS);
    return normalizeResult(responseJSON.Envelope[0].Body[0][soapAnnotations.elementName]);
  };
}

function createRootValue(schema: GraphQLSchema, fetchFn: MeshFetch) {
  const rootValue: Record<string, RootValueMethod> = {};
  const rootTypes = getRootTypes(schema);
  for (const rootType of rootTypes) {
    const rootFieldMap = rootType.getFields();
    for (const fieldName in rootFieldMap) {
      const annotations = getDirective(schema, rootFieldMap[fieldName], 'soap');
      const soapAnnotations: SoapAnnotations = Object.assign({}, ...annotations);
      rootValue[fieldName] = createRootValueMethod(soapAnnotations, fetchFn);
    }
  }
  return rootValue;
}

export function createExecutorFromSchemaAST(schema: GraphQLSchema, fetchFn: MeshFetch) {
  let rootValue: Record<string, RootValueMethod>;
  return function soapExecutor({ document, variables, context }) {
    if (!rootValue) {
      rootValue = createRootValue(schema, fetchFn);
    }
    return execute({
      schema,
      document,
      rootValue,
      contextValue: context,
      variableValues: variables,
      fieldResolver: defaultFieldResolver,
    });
  } as Executor;
}
