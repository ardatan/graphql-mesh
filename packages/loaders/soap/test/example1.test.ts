/* eslint-disable import/no-nodejs-modules */
import { SOAPLoader } from '../src';
import { promises } from 'fs';
import { printSchema } from 'graphql';
import { join } from 'path';
import { fetch } from '@whatwg-node/fetch';

const { readFile } = promises;

describe('Example 1', () => {
  it('should generate the schema correctly', async () => {
    const soapLoader = new SOAPLoader({
      fetch,
    });
    const example1Wsdl = await readFile(join(__dirname, './fixtures/example1.wsdl'), 'utf8');
    await soapLoader.loadWSDL(example1Wsdl);
    const schema = soapLoader.buildSchema();
    expect(printSchema(schema)).toMatchSnapshot();
  });
});
