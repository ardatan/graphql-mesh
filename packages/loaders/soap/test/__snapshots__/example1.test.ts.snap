// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example 1 should generate the schema correctly 1`] = `
"type Query {
  TodoService_TodoService_BasicHttpBinding_ITodoService_GetTodos(GetTodos: JSON = ""): TodoService_GetTodosResponse
}

type TodoService_GetTodosResponse {
  GetTodosResult: tns_ArrayOfTodoMessage!
}

type tns_ArrayOfTodoMessage {
  TodoMessage: [tns_TodoMessage]!
}

type tns_TodoMessage {
  Author: tns_Name!
  Description: String!
}

type tns_Name {
  First: String!
  Last: String!
  Middle: String!
}

"""
The \`JSON\` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSON

type Mutation {
  TodoService_TodoService_BasicHttpBinding_ITodoService_UpdateFileClass(UpdateFileClass: TodoService_UpdateFileClass_Input): JSON
  TodoService_TodoService_BasicHttpBinding_ITodoService_AddUpdateClass(AddUpdateClass: TodoService_AddUpdateClass_Input): TodoService_AddUpdateClassResponse
  TodoService_TodoService_BasicHttpBinding_ITodoService_AddTodo(AddTodo: TodoService_AddTodo_Input): TodoService_AddTodoResponse
  TodoService_TodoService_BasicHttpBinding_ITodoService_FaultAddTodo(FaultAddTodo: TodoService_FaultAddTodo_Input): TodoService_FaultAddTodoResponse
  TodoService_TodoService_BasicHttpBinding_ITodoService_FailGetTodo(FailGetTodo: JSON = ""): TodoService_FailGetTodoResponse
}

input TodoService_UpdateFileClass_Input {
  cls: tns_FileClass_Input
}

input tns_FileClass_Input {
  Mode: String
  ClassIntID: BigInt
  CreatedBy: String
  CreatedDtTm: DateTime
  FirmIntID: String
  LastUpdatedBy: String
  LastUpdatedDtTm: DateTime
  Name: String
  PeriodType: String
  Retention: Int
  Timestamp: Byte
}

"""
The \`BigInt\` scalar type represents non-fractional signed whole numeric values.
"""
scalar BigInt

"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar DateTime

"""The \`Byte\` scalar type represents byte value as a Buffer"""
scalar Byte

type TodoService_AddUpdateClassResponse {
  AddUpdateClassResult: Boolean!
}

input TodoService_AddUpdateClass_Input {
  updateClass: tns_UpdateClass_Input
}

input tns_UpdateClass_Input {
  ClassIndId: String
}

type TodoService_AddTodoResponse {
  AddTodoResult: Boolean!
}

input TodoService_AddTodo_Input {
  todo: tns_TodoMessage_Input
}

input tns_TodoMessage_Input {
  Author: tns_Name_Input
  Description: String
}

input tns_Name_Input {
  First: String
  Last: String
  Middle: String
}

type TodoService_FaultAddTodoResponse {
  FaultAddTodoResult: String!
}

input TodoService_FaultAddTodo_Input {
  text: String
}

type TodoService_FailGetTodoResponse {
  FailGetTodoResult: String!
}"
`;
